kernel/of/fdt.c

```diff
@@ -1,12 +1,340 @@
+
 #include <kernel/endian.h>
 #include <kernel/of/fdt.h>
 
+static int
+fdt_struct_ptr_check(struct fdt* fdt, void* ptr) {
+    void* start = ((void*)fdt) + be32toh(fdt->off_dt_struct);
+    void* end = start + be32toh(fdt->size_dt_struct);
+    if (ptr < start || ptr >= end) {
+        return 1;
+    }
+    return 0;
+}
+
+static uint32_t*
+fdt_token_begin(struct fdt* fdt) {
+    return (uint32_t*)(((void*)fdt) + be32toh(fdt->off_dt_struct));
+}
+
+static uint32_t*
+fdt_next_token(struct fdt* fdt, uint32_t* token) {
+    struct fdt_node* node;
+    struct fdt_prop* prop;
+    size_t str_len;
+    size_t cell_len;
+
+    uint32_t val = be32toh(*token);
+
+    switch (val) {
+        case FDT_NOP:
+        case FDT_END_NODE:
+            token++;
+            break;
+        case FDT_BEGIN_NODE:
+            node = (struct fdt_node*)token;
+            token++;
+            str_len = (size_t)strlen(node->unit_name) + 1;
+            cell_len = (str_len >> 2) + ((str_len & 3) != 0);
+            token += cell_len;
+            break;
+        case FDT_PROP:
+            prop = (struct fdt_prop*)token;
+            token++;
+            str_len = fdt_prop_val_len(prop);
+            cell_len = (str_len >> 2) + ((str_len & 3) != 0);
+            token += (2ull + cell_len);
+            break;
+        case FDT_END:
+        default:
+            token = NULL;
+            break;
+    }
+
+    if (fdt_struct_ptr_check(fdt, (void*)token)) {
+        return NULL;
```

Comment:
```suggestion
        return nullptr;
```

Reply:
Here there's literally no difference between NULL and nullptr afaik

Reply:
ones a builtin. It's just "more right"